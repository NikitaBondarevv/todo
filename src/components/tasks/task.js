import PropTypes from 'prop-types'

import styles from './styles.css'
import { getProgressClassName } from './getProgressClassName'
import { updateTask, deleteTask } from '../../contracts/tasks'

export const Task = ({ data, getTasks }) => {
  const { done, title } = data

  const completeTask = async () => {
    await updateTask({ ...data, done: true })
    getTasks()
  }
  
  const setInProgress = async () => {
    await updateTask({ ...data, done: undefined })
    getTasks()
  }

  const removeTask = async () => {
    await deleteTask(data)
    getTasks()
  }

  return (
    <li className={getProgressClassName(done)}>
      &nbsp;
      {title}
      {
        !done && (
          <div className={styles.buttons}>
            <button onClick={completeTask} type="button" />
            {
              done === undefined ? ''
                : (
                  <button onClick={setInProgress} className={styles.inProgress} type="button" />
                )
            }
            <button onClick={removeTask} className={styles.delete} type="button" />
          </div>
        )
      }
    </li>
  )
}

Task.defaultProps = {
  title: '',
  done: undefined,
}

Task.propTypes = {
  title: PropTypes.string,
  done: PropTypes.bool
}
