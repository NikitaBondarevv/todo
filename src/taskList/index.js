import { Component } from 'react'

import styles from './styles.css'

const Tasks = ({ id, completed, title, buttons }) => (
  <li key={id} className={completed ? styles.completed : styles.tasks}>
    <span className={completed ? styles.done : completed === undefined ? styles.progress : styles.waiting}></span>
    <p className={completed === undefined ? styles.taskInProgress : ''} >{`${id}. ${title}`}</p>
    {buttons}
  </li>
)

export class TaskList extends Component {
  state = {
    todos: [],
    value: ''
  }

  originTodos = []

  async getTasks() {
    const response = await fetch('https://jsonplaceholder.typicode.com/todos')
    const todos = await response.json()

    this.originTodos = todos

    this.setState({ todos: this.originTodos })
  }

  componentDidMount() {
    this.getTasks()
  }

  completeTask = (index) => {
    this.state.todos[index].completed = true

    this.setState({ todos: this.state.todos })
  }

  setInProgressTask = (index) => {
    this.state.todos[index].completed = undefined

    this.setState({ todos: this.state.todos })
  }

  deleteTask = (id) => {
    this.setState({
      todos: this.state.todos.filter(todo => todo.id !== id)
    })
  }

  taskSearch = ({ target: { value } }) => {
    this.setState({
      value,
      todos: this.originTodos
    })

    if (value.length > 1) {
      this.setState({ todos: this.originTodos.filter(todo => todo.title.includes(value)) })
    }
  }

  render() {
    const { todos, value } = this.state

    return (
      <>
        <input value={value} onChange={this.taskSearch} className={styles.search} placeholder="SEARCH" />
        <ul className={styles.tasksList}>
          {
            todos.map((todo, index) =>
              <Tasks key={index} id={index + 1} completed={todo.completed} title={todo.title} buttons={
                <div className={styles.buttons} >
                  <a onClick={() => this.completeTask(index)}></a>
                  {todo.completed === undefined ? '' : <a onClick={() => this.setInProgressTask(index)} className={styles.inProgress}></a>}
                  <a onClick={() => this.deleteTask(todo.id)} className={styles.delete}></a>
                </div>
              } />
            )
          }
        </ul>
      </>
    )
  }
}
