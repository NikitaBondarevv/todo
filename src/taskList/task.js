import styles from './styles.css'

export const Task = ({ index, done, title, buttons }) => (
  <li className={done ? styles.done : styles.tasks}>
    <span className={done ? styles.done : done === undefined ? styles.progress : styles.waiting}></span>
    <p className={done === undefined ? styles.taskInProgress : ''}>{`${index}. ${title}`}</p>
    {buttons}
  </li>
)
