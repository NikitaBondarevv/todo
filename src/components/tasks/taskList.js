import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import { UpdateCreateTask } from '../updateCreateTask'
import { Task } from './task'
import styles from './styles.css'

export const TaskList = ({ tasks, getTasks, activeTabIndex }) => {
  const [todos, setTodos] = useState(tasks)

  useEffect(() => {
    setTodos(tasks)
  }, [tasks])

  return (
    <>
      <div>
        <ol className={styles.tasksList}>
          {
            todos.map((task, index) => <Task key={index} data={task} getTasks={getTasks} />)
          }
        </ol>
        <a href="#" className={styles.addTask}>Add new task</a>
      </div>
      <UpdateCreateTask text="New Task" getTasks={getTasks} activeTabIndex={activeTabIndex} />
    </>
  )
}

TaskList.defaultProps = {
  tasks: [],
  getTasks: () => { },
  activeTabIndex: 0
}

TaskList.propTypes = {
  tasks: PropTypes.array,
  getTasks: PropTypes.func,
  activeTabIndex: PropTypes.number
}
