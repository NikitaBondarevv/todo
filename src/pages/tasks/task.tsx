import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { updateTask, deleteTask } from 'contracts/tasks'
import { getProgressClassName } from './getProgressClassName'
import { TTaskProps } from './types'
import styles from './styles.css'

export const Task = ({ data, getTasks }: TTaskProps) => {  
  const { done, title } = data

  const completeTask = async () => {
    await updateTask({ ...data, done: true })
    getTasks()
  }
  
  const setInProgress = async () => {
    await updateTask({ ...data, done: undefined })
    getTasks()
  }

  const removeTask = async () => {
    await deleteTask(data)
    getTasks()
  }

  return (
    <li className={getProgressClassName(done)}>
      &nbsp;
      <Link to={`/task/${data.day}/${data.id}`}>{title}</Link>
      {
        !done && (
          <div className={styles.buttons}>
            <button onClick={completeTask} type="button" />
            {
              done !== undefined && (
                  <button onClick={setInProgress} className={styles.inProgress} type="button" />
                )
            }
            <button onClick={removeTask} className={styles.delete} type="button" />
          </div>
        )
      }
    </li>
  )
}

Task.defaultProps = {
  data: {},
}

Task.propTypes = {
  data: PropTypes.object,
  getTasks: PropTypes.func
}
