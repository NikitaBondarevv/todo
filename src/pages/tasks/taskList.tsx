import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import { Task } from './task'
import { TTaskListProps } from './types'
import styles from './styles.css'

export const TaskList = ({ tasks, getTasks, activeTabIndex }: TTaskListProps) => {
  const [todos, setTodos] = useState(tasks)

  useEffect(() => {
    setTodos(tasks)
  }, [tasks])

  return (
    <div>
      <ol className={styles.tasksList}>
        {
          todos.map((task, index) => <Task key={index} data={task} getTasks={getTasks} />)
        }
      </ol>
      <Link to={`/task/${activeTabIndex}`} className={styles.addTask}>Add new task</Link>
    </div>
  )
}

TaskList.defaultProps = {
  tasks: [],
  activeTabIndex: 0,
  getTasks: () => {}
}

TaskList.propTypes = {
  tasks: PropTypes.array,
  getTasks: PropTypes.func,
  activeTabIndex: PropTypes.number
}
