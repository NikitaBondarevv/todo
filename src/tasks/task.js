import PropTypes from 'prop-types'

import styles from './styles.css'
import { getProgressClassName } from './getProgressClassName'
import { updateTask, deleteTask } from '../contracts/tasks'

export const Task = ({ data, getTasks }) => {
  const { done, title } = data

  const completeTask = async () => {
    await updateTask({ ...data, done: true })
    getTasks()
  }
  
  const setInProgress = async () => {
    await updateTask({ ...data, done: undefined })
    getTasks()
    console.log(data);
  }

  const deleteCurrentTask = async () => {
    await deleteTask(data)
    getTasks()
  }

  return (
    <li className={getProgressClassName(done)}>
      {title}
      {
        !done && (
          <div className={styles.buttons}>
            <a onClick={completeTask}>done</a>
            {
              done === undefined ? ''
                : (
                  <a onClick={setInProgress} className={styles.inProgress}>
                    in progress
                  </a>
                )
            }
            <a onClick={deleteCurrentTask} className={styles.delete}>delete</a>
          </div>
        )
      }
    </li>
  )
}

Task.defaultProps = {
  title: '',
  index: 0,
  done: undefined,
}

Task.propTypes = {
  title: PropTypes.string,
  index: PropTypes.number,
  done: PropTypes.bool
}
