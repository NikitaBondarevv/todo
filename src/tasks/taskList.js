import { Component } from 'react'
import PropTypes from 'prop-types'

import { Task } from './task'
import styles from './styles.css'

export class TaskList extends Component {
  static defaultProps = {
    tasks: []
  }

  static propTypes = {
    tasks: PropTypes.array
  }

  state = {
    todos: this.props.tasks,
    hidden: true,
    value: ''
  }

  componentDidUpdate(prevProps) {
    if (this.props.tasks !== prevProps.tasks) {

      this.setState({ todos: this.props.tasks })
    }
  }

  setValue = ({ target: { value } }) => {
    this.setState({ value })
  }

  showInput = () => {
    this.setState({ hidden: false })
  }

  randomInteger = () => {
    const rand = 1 + Math.random() * (100 + 1 - 1);

    return Math.floor(rand);
  }

  addTaskAndHandleBlur = (e) => {
    e.preventDefault()

    if (this.state.value.length > 2) {
      this.state.todos.push({
        title: this.state.value,
        id: this.randomInteger(),
        done: false
      })
    }

    this.setState({
      todos: this.state.todos,
      hidden: true,
      value: ''
    })
  }

  render() {
    const { todos, value, hidden } = this.state

    return (
      <div>
        <ul className={styles.tasksList}>
          {
            todos.map((task, index) => <Task key={index} index={index + 1} data={task} getTasks={this.props.getTasks} />)
          }
        </ul>
        <div className={styles.addTask}>
          {
            hidden ? <span onClick={this.showInput}>Add new task</span> :
              <form onSubmit={this.addTaskAndHandleBlur}>
                <input className={styles.editableText}
                  name="text"
                  value={value}
                  onChange={this.setValue}
                  onBlur={this.addTaskAndHandleBlur}
                  autoFocus
                />
              </form>
          }
        </div>
      </div>
    )
  }
}
