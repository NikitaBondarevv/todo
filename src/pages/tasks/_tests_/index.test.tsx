import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'

import { Tasks } from '..'
import { Task } from '../task'
import { TaskList } from '../taskList'

jest.mock('contracts/getTasks', () => ({
  getTasks: () => ({})
}))

jest.mock('components/tabs', () => ({
  Tabs:  () => <div />
}))

describe('<Tasks />, <Task />, <TaskList />', () => {
  test('should render div', () => {
    const { asFragment } = render(
      <Tasks />
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render li', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Task />
      </BrowserRouter>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render div', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <TaskList />
      </BrowserRouter>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
