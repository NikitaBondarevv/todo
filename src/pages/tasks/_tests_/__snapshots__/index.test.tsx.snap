// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tasks />, <Task />, <TaskList /> should render div 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`<Tasks />, <Task />, <TaskList /> should render div 2`] = `
<DocumentFragment>
  <div>
    <ol
      class="tasksList"
    />
    <a
      class="addTask"
      href="/task/0"
    >
      Add new task
    </a>
  </div>
</DocumentFragment>
`;

exports[`<Tasks />, <Task />, <TaskList /> should render li 1`] = `
<DocumentFragment>
  <li
    class="progress updateTask"
  >
    Â 
    <a
      href="/task/undefined/undefined"
    />
    <div
      class="buttons"
    >
      <button
        type="button"
      />
      <button
        class="delete"
        type="button"
      />
    </div>
  </li>
</DocumentFragment>
`;
